Enable `out_dims` for vmap frontend API

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #40717 Enable `in_dims` for vmap frontend api
* **#40576 Enable `out_dims` for vmap frontend API**
* #40575 Initial vmap docstring

`out_dims` specifies where in the output tensors the vmapped dimension
should appear. We implement this by simply creating a view with the
batch dimension moved to the desired position.

`out_dims` must either:
- be int (use the same value for all outputs)
- be Tuple[int] (so the user specifies one out_dim per output).
(See the vmap docstring for what we advertise out_dims to do).

I also renamed `TestVmap` to `TestVmapAPI` to make it clearer that we
are testing the API here and not specific operators (which will go into
their own test class).

Test Plan:
- `pytest test/test_vmap.py -v`

Differential Revision: [D22288086](https://our.internmc.facebook.com/intern/diff/D22288086)

zou3519 committed:
Enable `out_dims` for vmap frontend API

`out_dims` specifies where in the output tensors the vmapped dimension
should appear. We implement this by simply creating a view with the
batch dimension moved to the desired position.

`out_dims` must either:
- be int (use the same value for all outputs)
- be Tuple[int] (so the user specifies one out_dim per output).
(See the vmap docstring for what we advertise out_dims to do).

I also renamed `TestVmap` to `TestVmapAPI` to make it clearer that we
are testing the API here and not specific operators (which will go into
their own test class).

Test Plan:
- `pytest test/test_vmap.py -v`

[ghstack-poisoned]

ezyang commented:
Might take me a little time before I can review this one

ezyang commented on torch/_vmap_internals.py:
```diff
@@ -61,6 +93,13 @@ def _validate_outputs(outputs, fn_name):
+        return
+    if not isinstance(out_dims, tuple) or \
+            not all([isinstance(out_dim, int) for out_dim in out_dims]):
+        raise ValueError(OUT_DIMS_MUST_BE_INT_OR_TUPLE_OF_INT.format(out_dims=out_dims, fn=fn_name))
```
Now I am wondering, why the argument validation isn't done in the C binding code. A lot of this would be handled automatically (and more efficiently) by our existing Python binding code. I guess because there isn't a convenient "API" you can expose from C++ (but I wonder if this necessarily has to be the case...)

zou3519 replied:
Right now we accept `out_dims` as a tuple of ints. In the future, we'll want to accept `out_dims` as an arbitrary python nested data structure. It can have dictionaries, tuples, lists, etc nested together. Our C++ arg parser cannot handle something like that so we do it at the Python level as an initial implementation.

Furthermore, our C++ arg parser doesn't differentiate between "lists of int" and "tuples of int" (we only want to support tuple of int for now). There's a semantic difference between those two for vmap. `out_dims` should be a data structure that mirrors the data structure of the output of the function. In python, if a function returns multiple outputs, then it returns a tuple with said outputs. `out_dims` with a list of ints should be used in the case where a function returns a list of tensors.

ezyang commented on aten/src/ATen/native/Batching.cpp:
```diff
@@ -20,6 +21,66 @@ static bool has_level(const Tensor& self, int64_t level) {
 }
 
+// Returns a Tensor with batch dim with level `level` turned into a regular tensor,
+// as well as a logical dim index of where said dimension is in the returned tensor.
```
This is... kind of a weird thing to want to do.

zou3519 replied:
I can add in an additional comment about why we want to do this (I"ll just copy/paste the following)

The reason why we want to do this is because we want to keep track of which dimension used to be the batch dimension so that we can move it to the correct logical dimension. For example, if we had:
```
x = torch.randn(2, 3, 5)
vmap(lambda x: x, out_dims=1)(x)
```
then right when we are about to exist the vmap block, x is a BatchedTensor with a batch dimension at (physical) index 0. The batch dimension doesn't always have to exist at (physical) index 0.  When we undo the batch dimension, we want to move it to dimension 1 (as specified by `out_dims`). So we return the index at which the batch dim appears so that we can move it to the correct place.

zou3519 replied:
Hmm, I realize that the comment is wrong the way it is written and that that might be the source of the confusion. Let me fix.

ezyang replied:
Is there ever a situation where you'd want to this operation without immediately moving the batch dimension afterwards?

zou3519 replied:
No, there isn't -- would it be easier to read if I tossed the `movedim` calls into this function?

ezyang replied:
Not necessarily, the function is pretty long. The difficulty is distinguishing between whether or not this is a semantically meaningful operation, or if it's just a helper operation for some other semantically meaningful operation. C++ doesn't make it too easy to express your intent here.

zou3519 committed:
Update on "Enable `out_dims` for vmap frontend API"


`out_dims` specifies where in the output tensors the vmapped dimension
should appear. We implement this by simply creating a view with the
batch dimension moved to the desired position.

`out_dims` must either:
- be int (use the same value for all outputs)
- be Tuple[int] (so the user specifies one out_dim per output).
(See the vmap docstring for what we advertise out_dims to do).

I also renamed `TestVmap` to `TestVmapAPI` to make it clearer that we
are testing the API here and not specific operators (which will go into
their own test class).

Test Plan:
- `pytest test/test_vmap.py -v`

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/Batching.cpp:
```diff
@@ -20,6 +21,73 @@ static bool has_level(const Tensor& self, int64_t level) {
+// index at which the batch dim appears so that we can move it to the correct place.
+// later down the line.
+static std::pair<Tensor,int64_t> remove_existing_batch_dim(const Tensor& self, int64_t level) {
+  const auto* batched = maybeGetBatched(self);
```
minor nit: consider passing in `batched` as the argument, it lets you statically enforce the invariant, rather than a runtime test.

zou3519 replied:
Changed to take `batched` as argument

ezyang commented on aten/src/ATen/native/Batching.cpp:
```diff
@@ -20,6 +21,73 @@ static bool has_level(const Tensor& self, int64_t level) {
+    return std::make_pair(batched->value(), bdims[0].dim());
+  }
+  BatchDims new_bdims;
+  int64_t newly_exposed_physical_dim;
```
It seems possible for `newly_exposed_physical_dim` to pass through the loop below without being set. I guess there is a bdims invariant which would ensure this is the case. But you could be extra careful by initializing this to -1 and then asserting that it is not -1 at the end.

zou3519 replied:
The assert sounds good. Yeah, this function assumes that the level must exist in the bdims.

zou3519 replied:
Added assert

ezyang commented on aten/src/ATen/native/Batching.cpp:
```diff
@@ -20,6 +21,73 @@ static bool has_level(const Tensor& self, int64_t level) {
+  auto bdims = batched->bdims();
+  if (bdims.size() == 1) {
+    TORCH_INTERNAL_ASSERT(bdims[0].level() == level);
+    return std::make_pair(batched->value(), bdims[0].dim());
```
Crumbs, I've gotten confused again. Is the dim() of a BatchDim logical or physical? I read the comment but it didn't help me understand:

```
// a BatchDim represents a "private" dimension on a Tensor created inside of
// vmap. It is a (level, dim) tuple, with the `dim` indicating which dimension
// is being vmap'ed over and the `level` being an identifier for which vmap
// said dimension was created inside.
```

ezyang replied:
I guess, based on 

```
// The batch dimensions are treated as being "private"; they are not user-visible.
// For example, in the following Tensor,
//    bt = BatchedTensorImpl(ones(2, 3, 5, 7), [(lvl=1, dim=0), (lvl=2, dim=1)])
// dimensions 0 and 1 are batch dimensions.

```

this means the dim is physical.

ezyang replied:
OK, so in this case it's ok to pun the physical dim into a logical dim because you've eliminated all of the batch dimensions in this case. :+1:

zou3519 replied:
dim() of a BatchDim is physical. I'll go beef up the documentation for BatchDim

zou3519 replied:
Added a note to `BatchDim` that the dimension index is physical

ezyang commented on aten/src/ATen/native/Batching.cpp:
```diff
@@ -20,6 +21,73 @@ static bool has_level(const Tensor& self, int64_t level) {
+// always have to exist at (physical) index 0. When we undo the batch dimension,
+// we want to move it to dimension 1 (as specified by out_dims). So we return the
+// index at which the batch dim appears so that we can move it to the correct place.
+// later down the line.
```
Is it valid to specify a level that doesn't exist in the tensor?

ezyang replied:
Judging from

```
    TORCH_INTERNAL_ASSERT(bdims[0].level() == level);
```

sounds like the answer is no

zou3519 replied:
It is indeed not valid to specify a level that doesn't exist in the tensor. The `_remove_batch_dim` function that calls `remove_existing_batch_dim` handles the case where the level doesn't exist in the tensor separately

zou3519 replied:
Added a line to the comment that the level must exist in the tensor.

ezyang commented on aten/src/ATen/native/Batching.cpp:
```diff
@@ -20,6 +21,73 @@ static bool has_level(const Tensor& self, int64_t level) {
+    }
+    permutation.push_back(dim);
+  }
+  permutation.insert(permutation.begin() + dst, src);
```
O(n) police! (Should be fine)

zou3519 replied:
If you saw the implementation of np.moveaxis... there would be an O(n^2) police sounding alarms. Otherwise I would have just ported np.moveaxis in PyTorch and used it here

ezyang reviewed (approved):
Logic looks correct, comments mostly for clarity.

zou3519 committed:
Update on "Enable `out_dims` for vmap frontend API"


`out_dims` specifies where in the output tensors the vmapped dimension
should appear. We implement this by simply creating a view with the
batch dimension moved to the desired position.

`out_dims` must either:
- be int (use the same value for all outputs)
- be Tuple[int] (so the user specifies one out_dim per output).
(See the vmap docstring for what we advertise out_dims to do).

I also renamed `TestVmap` to `TestVmapAPI` to make it clearer that we
are testing the API here and not specific operators (which will go into
their own test class).

Test Plan:
- `pytest test/test_vmap.py -v`

[ghstack-poisoned]
